### PR TITLE
feat(build): add GitHub Action to make release

### PR DESCRIPTION
**Pull request by PsypherPunk**
Add config. to, when a new tag is pushed:

- build a `.deb` file using [`cargo-deb`](https://github.com/kornelski/cargo-deb);
- include the resulting `.deb` file in the subsequent release.

An example of one such release can be seen [here](https://github.com/PsypherPunk/RustScan/releases/tag/v6.6.6).

Relates #83

### CODE REVIEW

**Comment by bee-san** [comment]
Merge conflicts :(

**Comment by PsypherPunk** [comment]
@bee-san, I can look to resolve these but is there still interest in this PR?

**Comment by bee-san** [comment]
> @bee-san, I can look to resolve these but is there still interest in this PR?

Oooh great question!

We have this release.yaml https://github.com/RustScan/RustScan/blob/master/.github/workflows/release.yml

But it does not do .deb , and we'd like to make it do .deb 😅 

I am thinking you can just copy this PR's yaml into that one?

**Comment by PsypherPunk** [comment]
@bee-san, sure, I'll take a look. There's some nice matrix-build steps going on in there so I'll try to do something with it.

**Comment by bee-san** [comment]
> @bee-san, sure, I'll take a look. There's some nice matrix-build steps going on in there so I'll try to do something with it.

Thank you, that would honestly be awesome. Before this we had a bulky Docker-builder to set up debs (feel free to delete https://github.com/RustScan/RustScan/tree/master/rustscan-debbuilder in your PR) 🙈 

Because of how cumbersome it was, new releases don't have .deb files :( 

I was just saying today one of the most important things I should add is .deb building automation!!

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -121,6 +128,11 @@ jobs:
           else
             cp "target/${{ matrix.target }}/release/$BIN_NAME" "dist/"
           fi
+
+          if [ "${{ matrix.build }}" = "x86_64-linux" ]; then
+            find target/ -name "*.deb"
+            cp "$(find target/ -name "*.deb")" "dist/"
```
Comment (PsypherPunk) [inline .github/workflows/release.yml:134]: This is _odd_.

You can see in [this](https://github.com/PsypherPunk/RustScan/actions/runs/8598981577/job/23560927701) run that the `find` here results in `target/x86_64-unknown-linux-gnu/debian/rustscan_2.1.1-1_amd64.deb`.

However, adding `cp target/x86_64-unknown-linux-gnu/debian/rustscan_*-1_amd64.deb` results in a `cp: cannot stat` error despite the `find` showing that it does exist…?!

**Comment by PsypherPunk** [comment]
@bee-san, PR updated.

I've moved it to draft in case you have feedback; there's on odd line to which I've added a comment; it works—you can see the resulting release [here](https://github.com/PsypherPunk/RustScan/releases/tag/v6.6.6)—but there's probably a much better way to handle it (I could pull out the version as is done later and resolve the path to the `.deb` complete, I suppose).

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -112,6 +112,13 @@ jobs:
             rustembedded/cross:${{ matrix.target }} \
             aarch64-linux-gnu-strip \
             /target/${{ matrix.target }}/release/$BIN_NAME
+
+      - name: Build deb package (linux only)
+        if: matrix.build == 'x86_64-linux'
+        run: |
+          cargo install cargo-deb
+          cargo deb --target ${{ matrix.target }}
```
Comment (PsypherPunk) [inline .github/workflows/release.yml:120]: The `--deb-version` might be wiser here: elsewhere, versions are derived from the _tag_, not the Cargo version.

Comment (PsypherPunk) [reply .github/workflows/release.yml]: Using `--deb-version` (which makes the path a little more deterministic too).

**Comment by bee-san** [comment]
want me to review? looks ok to you? :)

**Comment by PsypherPunk** [comment]
@bee-san, it works insofar as it generate a `.deb` file and attaches it to a release as per the existing `release.yml` so…yes 😬

As long as you're happy with the changes (and I deleted the right files you mentioned) then LGTM.

**Comment by bee-san** [comment]
> @bee-san, it works insofar as it generate a `.deb` file and attaches it to a release as per the existing `release.yml` so…yes 😬
> 
> As long as you're happy with the changes (and I deleted the right files you mentioned) then LGTM.

Thanks!!

We only build for x86_64-linux? Do you think we should use the matrix to add support for more architectures or? 🤔

**Comment by PsypherPunk** [comment]
I'm assuming it would only make sense to build `.deb` files for Linux and the only arch. is currently `x86_64-linux`…?

If that changes (say, adding `i686`), it should just be a matter of changing those `== 'x86_64-linux'` checks to `=~ 'linux'` or similar.

However, I'd probably consider doing that in a separate PR, once this one's proven with the existing arch./release-process…?

**Comment by bee-san** [comment]
> I'm assuming it would only make sense to build `.deb` files for Linux and the only arch. is currently `x86_64-linux`…?
> 
> If that changes (say, adding `i686`), it should just be a matter of changing those `== 'x86_64-linux'` checks to `=~ 'linux'` or similar.
> 
> However, I'd probably consider doing that in a separate PR, once this one's proven with the existing arch./release-process…?

Ok thank you! GitHub's matrix sucks :P

**Comment by bee-san** [comment]
It appears the release did not run? :( 

https://github.com/PsypherPunk/RustScan/actions/workflows/release.yml

```
name: Release
on:
  push:
    tags:
      - '[0-9]+.[0-9]+.[0-9]+'
  pull_request:
    branches:
      - master
    paths:
      - 'Cargo.toml'
      - 'Cargo.lock'
      - '.github/workflows/release.yml'
```

The latest tag is `2.2.2`, I am nto sure why it would not run :( 

https://github.com/RustScan/RustScan/releases/tag/2.2.2

**Comment by PsypherPunk** [comment]
@bee-san, can I check how you actually created the tag? 

Not entirely sure if/why that would matter but I'll take a look 😬

**Comment by bee-san** [comment]
> @bee-san, can I check how you actually created the tag?
> 
> Not entirely sure if/why that would matter but I'll take a look 😬

I made it in GitHub releases with 
image

thank you for your contributions!

**Comment by bee-san** [comment]
image

Ok so it did run eventually, that's nice to know!! Docker and homebrew did not like it, stuff to fix 😂

**Comment by PsypherPunk** [comment]
Phew! And there's a `.deb` an' everything!

**Pull request merged**